src/core/server.cc

```diff
@@ -294,150 +293,354 @@ class HTTPServiceImpl {
+void
+HTTPServiceImpl::thread_init(evhtp_t* htp, evthr_t* thread, void* arg)
+{
+  evthr_set_aux(thread, new ThreadWiseService(thread));
+}
+
+void
+HTTPServiceImpl::thread_exit(evhtp_t* htp, evthr_t* thread, void* arg)
+{
+  ThreadWiseService* requests = (ThreadWiseService*)evthr_get_aux(thread);
+  delete requests;
+}
+
+tensorflow::Status
+HTTPServiceImpl::Start(uint16_t port, int thread_cnt)
+{
+  if (!worker_.joinable()) {
+    evbase_ = event_base_new();
+    htp_ = evhtp_new(evbase_, NULL);
+    evhtp_set_gencb(htp_, HTTPServiceImpl::Dispatch, this);
+    evhtp_use_threads_wexit(htp_, thread_init, thread_exit, thread_cnt, NULL);
+    evhtp_bind_socket(htp_, "0.0.0.0", port, 1024);
+    worker_ = std::thread(event_base_loop, evbase_, 0);
+    return tensorflow::Status::OK();
+  }
+
+  return tensorflow::Status(
+      tensorflow::error::ALREADY_EXISTS, "HTTP server is already running.");
+}
+
+tensorflow::Status
+HTTPServiceImpl::Stop()
+{
+  if (worker_.joinable()) {
+    event_base_loopexit(evbase_, NULL);
+    worker_.join();
+    return tensorflow::Status::OK();
+  }
+
+  return tensorflow::Status(
+      tensorflow::error::UNAVAILABLE, "HTTP server is not running.");
+}
+
+std::string
+HTTPServiceImpl::ToMethodString(htp_method method)
 {
-  return
-      [this](tfs::net_http::ServerRequestInterface* req) { this->Handle(req); };
+  switch (method) {
+    case htp_method_GET:
+      return "GET";
+    case htp_method_POST:
+      return "POST";
+    case htp_method_PUT:
+      return "PUT";
+    case htp_method_DELETE:
+      return "DELETE";
+    default:
+      return "UNKNOWN: " + std::to_string(method);
+  }
 }
 
 void
-HTTPServiceImpl::Handle(tfs::net_http::ServerRequestInterface* req)
+HTTPServiceImpl::Dispatch(evhtp_request_t* req, void* arg)
 {
-  LOG_VERBOSE(1) << "HTTP request: " << req->http_method() << " "
-                 << req->uri_path();
+  (static_cast<HTTPServiceImpl*>(arg))->Handle(req);
+}
 
-  tfs::net_http::HTTPStatusCode status =
-      tfs::net_http::HTTPStatusCode::BAD_REQUEST;
+void
+HTTPServiceImpl::Handle(evhtp_request_t* req)
+{
+  LOG_VERBOSE(1) << "HTTP request: " << ToMethodString(req->method) << " "
+                 << req->uri->path->full;
 
   std::string endpoint, rest;
   if (RE2::FullMatch(
-          std::string(req->uri_path()), api_regex_, &endpoint, &rest)) {
+          std::string(req->uri->path->full), api_regex_, &endpoint, &rest)) {
     // health
     if (endpoint == "health") {
-      status = Health(req, rest);
+      Health(req, rest);
     }
     // profile
     else if (endpoint == "profile") {
-      status = Profile(req, rest);
+      Profile(req, rest);
     }
     // infer
     else if (endpoint == "infer") {
-      status = Infer(req, rest);
+      Infer(req, rest);
     }
     // status
     else if (endpoint == "status") {
-      status = Status(req, rest);
+      Status(req, rest);
     }
   }
-
-  if (status != tfs::net_http::HTTPStatusCode::OK) {
```

Comment:
What happens to the request if it is not one of the above? Don't you need to response with error status? Isn't there a test for this case? If not you should add something to L0_lifecycle that makes a request for an invalid endpoint.